### ISSUE TITLE
0.25.2 generates a 409/Conflict responses with Async results

### ISSUE THREAD

**Issue by aheritier** [spring-projects/spring-hateoas#1096]
Hi team,

  I discovered a regression after upgrading from Spring Boot 2.1.6 to 2.1.7 and especially with the upgrade from hateoas 0.25.1 to 0.25.2.

  We have a controller defined like this (note the async result)

```
@RestController
@RequestMapping("/api/foo")
@ExposesResourceFor(FooStuffs.class)
public class FooStuffsController {

  @GetMapping(produces = MediaType.APPLICATION_JSON_VALUE)
  public Future<FooStuffs> getFooFromToDate(@RequestParam("fromDate") @DateTimeFormat(iso = DateTimeFormat.ISO.DATE)  LocalDate fromDate,
                                               @RequestParam(value = "toDate", required = false) @DateTimeFormat(iso = DateTimeFormat.ISO.DATE)  LocalDate optionalToDate) {
....
}
```

Which was working before:

```
2019-10-18 21:51:54.386 DEBUG 7242 --- [  XNIO-1 task-3] o.s.web.servlet.DispatcherServlet        : "ASYNC" dispatch for GET "/api/foo?fromDate=2019-10-01&toDate=2019-10-18", parameters={masked}
2019-10-18 21:51:54.390 DEBUG 7242 --- [  XNIO-1 task-3] s.w.s.m.m.a.RequestMappingHandlerAdapter : Resume with async result [FooStuffs{fromDate=2019-10-01, toDate=2019-10-08, foo=[]}]
```

And now it fails with a Conflict / 409 http error:

```
2019-10-18 21:55:30.858 DEBUG 7552 --- [  XNIO-1 task-3] o.s.web.servlet.DispatcherServlet        : "ASYNC" dispatch for GET "/api/foo?fromDate=2019-10-01&toDate=2019-10-18", parameters={masked}
2019-10-18 21:55:30.861 DEBUG 7552 --- [  XNIO-1 task-3] s.w.s.m.m.a.RequestMappingHandlerAdapter : Resume with async result [java.lang.IllegalStateException: Cannot ask for request attribute - request is not active anymore!]
```

Digging a bit (thanks @bclozel @snicoll) I found that it is a side effect of PR #749 

The problem is about the cache which is using `RequestAttributes.SCOPE_REQUEST`

https://github.com/spring-projects/spring-hateoas/commit/77392de587afc2b02b300a5129b9146af952adf9#diff-98e861a165d7d9d2bc81e5ecd48cdb96R114-R125

but it's not active because of the async and thus we get the `IllegalStateException`

https://github.com/spring-projects/spring-framework/blob/5.1.x/spring-web/src/main/java/org/springframework/web/context/request/ServletRequestAttributes.java#L149

I am not sure if hateoas is supposed to support Async results (I didn't find any doc to deny or validate it). Let me know if I can help more @odrotbohm @gregturn 

NOTE: I didn't try with 1.x but as far as I can see the code of  `UriComponentsBuilderFactory` didn't change.

**Comment by aheritier** [comment]
I will try to come back with a sample. The problem is still here with hateoas 1.0.3 (Spring Boot 2.2.5)

**Comment by gregturn** [comment]
Seeing as we have no test cases surrounding async Spring MVC behavior, I'd like to work on one so we have somewhere to go. First and foremost, I want to see if this issue exists on the `master` branch, and then work from there. 

If it does, we patch it and then assess backporting the fix to earlier versions.

**Comment by aheritier** [comment]
I still had this task to create a sample for you on my todo but I didn't progress.
I was thinking about writing directly some tests but I didn't have the time to check how they are written and if they could be hard to create.
As I didn't write this part of the code in our app I still have to check if this is really making sense to use such async with Hateoas (even if it's not forbidden ... it should work but we all know between the theory and the reality). It was working in < 0.25.1 doesn't mean that it makes sense.
I just hope that this bug isn't due to the combination of more things than just Hateoas + Async
If that's the case I will have to verify what other element could interfere.

**Comment by gregturn** [comment]
For starters, this fragment worked just fine, out-of-the-box against both Spring HATEAOS 1.0.4.RELEASE and 1.1.0.RC1:

```java
	@GetMapping("/employees")
	public DeferredResult<ResponseEntity<CollectionModel<EntityModel<Employee>>>> findAll() {

		DeferredResult<ResponseEntity<CollectionModel<EntityModel<Employee>>>> deferredResult = new DeferredResult<>(60L);

		ResponseEntity<CollectionModel<EntityModel<Employee>>> response = ResponseEntity.ok( //
				this.assembler.toCollectionModel(this.repository.findAll()));

		deferredResult.setResult(response);

		return deferredResult;

	}
```

**Comment by gregturn** [comment]
This also worked:

```java
@Async
@GetMapping("/employees")
public CompletableFuture<ResponseEntity<CollectionModel<EntityModel<Employee>>>> findAll() throws InterruptedException {

	ResponseEntity<CollectionModel<EntityModel<Employee>>> response = ResponseEntity.ok( //
			this.assembler.toCollectionModel(this.repository.findAll()));

	Thread.sleep(1000);

	return CompletableFuture.completedFuture(response);

}
```

**Comment by gregturn** [comment]
Changing the return type to `Future` just yielded:

```
{
  "cancelled": false,
  "done": false
}
```

**Comment by gregturn** [comment]
For the record, I have used Spring HATEOAS Examples to do all this, and thus, can't turn the dials back to 0.25.2 (via Spring Boot 2.1.13) and get this example to work.

**Comment by aheritier** [comment]
Thanks for your investigations @gregturn 
I am doing few more tests on my side and using CompletableFuture instead of Future does not help

I am almost sure that `Cannot ask for request attribute - request is not active anymore!` is related to something wrongly configured in my application at the Async level (but I don't find anything relevant for now) and potentially related to Spring Security but it's ... strange

**Comment by gregturn** [comment]
I'm glad you've identified that it's not a `0.25.x` issue, specifically.

That error message is found in Spring Framework inside `ServletRequestAttributes.getAttribute()`.

It is called directly by Spring HATEOAS's `UriComponentsBuilderFactory.getCachedBaseUri()`. It seems to hinge on the deferred result no longer being active. Smells like something has exceeded a timeout threshold. 

Thoughts @rstoyanchev?

**Comment by aheritier** [comment]
@gregturn 0.25.x is the root cause because downgrading it solves the issue.
There is a regression but I am not sure if hateoas is doing something wrong if our app is doing something not supported.
I am using my day off today to build a sample app to reproduce the issue.

**Comment by aheritier** [comment]
The full stacktrace I get with 0.25.2+ is

````
2020-05-06 16:39:31.603  INFO 81289 --- [  XNIO-1 task-1] io.undertow.servlet                      : Initializing Spring DispatcherServlet 'dispatcherServletRegistration'
2020-05-06 16:39:32.579 ERROR 81289 --- [  XNIO-1 task-2] io.undertow.request                      : UT005023: Exception handling request to /api/results

org.springframework.web.util.NestedServletException: Request processing failed; nested exception is java.lang.IllegalStateException: Cannot ask for request attribute - request is not active anymore!
	at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:1014)
	at org.springframework.web.servlet.FrameworkServlet.doGet(FrameworkServlet.java:898)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:645)
	at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:883)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:750)
	at io.undertow.servlet.handlers.ServletHandler.handleRequest(ServletHandler.java:74)
	at io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter(FilterHandler.java:129)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:320)
	at org.springframework.security.web.access.intercept.FilterSecurityInterceptor.invoke(FilterSecurityInterceptor.java:116)
	at org.springframework.security.web.access.intercept.FilterSecurityInterceptor.doFilter(FilterSecurityInterceptor.java:91)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
	at org.springframework.security.web.access.ExceptionTranslationFilter.doFilter(ExceptionTranslationFilter.java:119)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
	at org.springframework.security.web.session.SessionManagementFilter.doFilter(SessionManagementFilter.java:84)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
	at org.springframework.security.web.authentication.AnonymousAuthenticationFilter.doFilter(AnonymousAuthenticationFilter.java:111)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
	at org.springframework.security.web.servletapi.SecurityContextHolderAwareRequestFilter.doFilter(SecurityContextHolderAwareRequestFilter.java:170)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
	at org.springframework.security.web.savedrequest.RequestCacheAwareFilter.doFilter(RequestCacheAwareFilter.java:63)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:103)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
	at org.springframework.security.web.authentication.logout.LogoutFilter.doFilter(LogoutFilter.java:116)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:103)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
	at org.springframework.security.web.context.SecurityContextPersistenceFilter.doFilter(SecurityContextPersistenceFilter.java:105)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:103)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
	at org.springframework.security.web.FilterChainProxy.doFilterInternal(FilterChainProxy.java:215)
	at org.springframework.security.web.FilterChainProxy.doFilter(FilterChainProxy.java:178)
	at org.springframework.web.filter.DelegatingFilterProxy.invokeDelegate(DelegatingFilterProxy.java:358)
	at org.springframework.web.filter.DelegatingFilterProxy.doFilter(DelegatingFilterProxy.java:271)
	at io.undertow.servlet.core.ManagedFilter.doFilter(ManagedFilter.java:61)
	at io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter(FilterHandler.java:131)
	at org.springframework.boot.actuate.metrics.web.servlet.WebMvcMetricsFilter.filterAndRecordMetrics(WebMvcMetricsFilter.java:114)
	at org.springframework.boot.actuate.metrics.web.servlet.WebMvcMetricsFilter.doFilterInternal(WebMvcMetricsFilter.java:104)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:119)
	at io.undertow.servlet.core.ManagedFilter.doFilter(ManagedFilter.java:61)
	at io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter(FilterHandler.java:131)
	at io.undertow.servlet.handlers.FilterHandler.handleRequest(FilterHandler.java:84)
	at io.undertow.servlet.handlers.security.ServletSecurityRoleHandler.handleRequest(ServletSecurityRoleHandler.java:62)
	at io.undertow.servlet.handlers.ServletChain$1.handleRequest(ServletChain.java:68)
	at io.undertow.servlet.handlers.ServletDispatchingHandler.handleRequest(ServletDispatchingHandler.java:36)
	at io.undertow.servlet.handlers.RedirectDirHandler.handleRequest(RedirectDirHandler.java:68)
	at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
	at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
	at io.undertow.servlet.handlers.ServletInitialHandler.handleFirstRequest(ServletInitialHandler.java:269)
	at io.undertow.servlet.handlers.ServletInitialHandler.access$100(ServletInitialHandler.java:78)
	at io.undertow.servlet.handlers.ServletInitialHandler$2.call(ServletInitialHandler.java:133)
	at io.undertow.servlet.handlers.ServletInitialHandler$2.call(ServletInitialHandler.java:130)
	at io.undertow.servlet.core.ServletRequestContextThreadSetupAction$1.call(ServletRequestContextThreadSetupAction.java:48)
	at io.undertow.servlet.core.ContextClassLoaderSetupAction$1.call(ContextClassLoaderSetupAction.java:43)
	at io.undertow.servlet.handlers.ServletInitialHandler.dispatchRequest(ServletInitialHandler.java:249)
	at io.undertow.servlet.handlers.ServletInitialHandler.dispatchToPath(ServletInitialHandler.java:186)
	at io.undertow.servlet.spec.AsyncContextImpl$2$1.handleRequest(AsyncContextImpl.java:194)
	at io.undertow.server.Connectors.executeRootHandler(Connectors.java:376)
	at io.undertow.servlet.spec.AsyncContextImpl$2.run(AsyncContextImpl.java:191)
	at io.undertow.servlet.spec.AsyncContextImpl$6.run(AsyncContextImpl.java:465)
	at io.undertow.servlet.spec.AsyncContextImpl$TaskDispatchRunnable.run(AsyncContextImpl.java:584)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: java.lang.IllegalStateException: Cannot ask for request attribute - request is not active anymore!
	at org.springframework.web.context.request.ServletRequestAttributes.getAttribute(ServletRequestAttributes.java:149)
	at org.springframework.hateoas.mvc.UriComponentsBuilderFactory.getCachedBaseUri(UriComponentsBuilderFactory.java:124)
	at org.springframework.hateoas.mvc.UriComponentsBuilderFactory.getBuilder(UriComponentsBuilderFactory.java:57)
	at org.springframework.hateoas.mvc.ControllerLinkBuilderFactory.linkTo(ControllerLinkBuilderFactory.java:141)
	at org.springframework.hateoas.mvc.ControllerLinkBuilder.linkTo(ControllerLinkBuilder.java:182)
	at com.example.demo.rest.UsersResultsController$UsersResultsLinkBuilder.withSelfLink(UsersResultsController.java:83)
	at com.example.demo.rest.UsersResultsController.lambda$getResultsFromToDate$3(UsersResultsController.java:70)
	at java.base/java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:642)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:506)
	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1705)
	... 3 common frames omitted
```

**Comment by gregturn** [comment]
Hmm. Undertow is another variable I hadn't considered. What happens if you switch servlet containers?

**Comment by aheritier** [comment]
Same with Tomcat @gregturn 

```
May 06, 2020 5:33:28 PM org.apache.catalina.core.ApplicationContext log
INFO: Initializing Spring DispatcherServlet 'dispatcherServletRegistration'
May 06, 2020 5:33:37 PM org.apache.catalina.core.ApplicationDispatcher invoke
SEVERE: Servlet.service() for servlet [dispatcherServletRegistration] threw exception
java.lang.IllegalStateException: Cannot ask for request attribute - request is not active anymore!
	at org.springframework.web.context.request.ServletRequestAttributes.getAttribute(ServletRequestAttributes.java:149)
	at org.springframework.hateoas.mvc.UriComponentsBuilderFactory.getCachedBaseUri(UriComponentsBuilderFactory.java:124)
	at org.springframework.hateoas.mvc.UriComponentsBuilderFactory.getBuilder(UriComponentsBuilderFactory.java:57)
	at org.springframework.hateoas.mvc.ControllerLinkBuilderFactory.linkTo(ControllerLinkBuilderFactory.java:141)
	at org.springframework.hateoas.mvc.ControllerLinkBuilder.linkTo(ControllerLinkBuilder.java:182)
	at com.example.demo.rest.UsersActivitiesController$UsersActivitiesLinkBuilder.withSelfLink(UsersActivitiesController.java:85)
	at com.example.demo.rest.UsersActivitiesController.lambda$getUsersFromToDate$3(UsersActivitiesController.java:71)
	at java.base/java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:642)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:506)
	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1705)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)

May 06, 2020 5:33:37 PM org.apache.catalina.core.StandardWrapperValve invoke
SEVERE: Servlet.service() for servlet [dispatcherServletRegistration] in context with path [] threw exception [Request processing failed; nested exception is java.lang.IllegalStateException: Cannot ask for request attribute - request is not active anymore!] with root cause
java.lang.IllegalStateException: Cannot ask for request attribute - request is not active anymore!
	at org.springframework.web.context.request.ServletRequestAttributes.getAttribute(ServletRequestAttributes.java:149)
	at org.springframework.hateoas.mvc.UriComponentsBuilderFactory.getCachedBaseUri(UriComponentsBuilderFactory.java:124)
	at org.springframework.hateoas.mvc.UriComponentsBuilderFactory.getBuilder(UriComponentsBuilderFactory.java:57)
	at org.springframework.hateoas.mvc.ControllerLinkBuilderFactory.linkTo(ControllerLinkBuilderFactory.java:141)
	at org.springframework.hateoas.mvc.ControllerLinkBuilder.linkTo(ControllerLinkBuilder.java:182)
	at com.example.demo.rest.UsersActivitiesController$UsersActivitiesLinkBuilder.withSelfLink(UsersActivitiesController.java:85)
	at com.example.demo.rest.UsersActivitiesController.lambda$getUsersFromToDate$3(UsersActivitiesController.java:71)
	at java.base/java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:642)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:506)
	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1705)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
```

Still trying to minimize the app as much as possible but it's not easy.

**Comment by aheritier** [comment]
@gregturn I found the issue \o/ and I am happy to share that it's not bug in hateoas !!!

I found that a previous developer implemented a workaround to allow hateoas to work with Async when it was not supported.
For this he wrote a specific config for the Servlet Dispatcher .... 
While it was explained in the class comment the name of the Config amongst all others never caught my eye. Removing it is solving my issue. The bug was in our app because of the conflict between the workaround and the fix done in 0.25.2 ...
I am now safe to upgrade

```
/**
 * Fixes issue in Spring HATEOAS where RequestContextHolder is not available when using asynchronous service requests.
 */
@Configuration
@ConditionalOnWebApplication
@EnableConfigurationProperties(WebMvcProperties.class)
public class DispatcherServletConfig {

  private final WebMvcProperties webMvcProperties;

  public DispatcherServletConfig(WebMvcProperties webMvcProperties) {
    this.webMvcProperties = webMvcProperties;
  }

  @Bean(name = DEFAULT_DISPATCHER_SERVLET_BEAN_NAME)
  public DispatcherServlet dispatcherServlet() {
    DispatcherServlet dispatcherServlet = new DispatcherServlet();

    dispatcherServlet.setDispatchOptionsRequest(this.webMvcProperties.isDispatchOptionsRequest());
    dispatcherServlet.setDispatchTraceRequest(this.webMvcProperties.isDispatchTraceRequest());
    dispatcherServlet.setThrowExceptionIfNoHandlerFound(this.webMvcProperties.isThrowExceptionIfNoHandlerFound());

    return dispatcherServlet;
  }

  @Bean(name = DEFAULT_DISPATCHER_SERVLET_REGISTRATION_BEAN_NAME)
  public ServletRegistrationBean<DispatcherServlet> dispatcherServletRegistration(DispatcherServlet dispatcherServlet) {
    ServletRegistrationBean<DispatcherServlet> registration = new ServletRegistrationBean<>(dispatcherServlet, this.webMvcProperties.getServlet().getServletMapping());

    registration.setName(DEFAULT_DISPATCHER_SERVLET_REGISTRATION_BEAN_NAME);
    registration.setLoadOnStartup(this.webMvcProperties.getServlet().getLoadOnStartup());

    dispatcherServlet.setThreadContextInheritable(true); // Enable request attributes in spawned threads

    return registration;
  }

  @Bean
  @ConditionalOnMissingBean(DispatcherServletPath.class)
  @ConditionalOnSingleCandidate(DispatcherServlet.class)
  public DispatcherServletPath dispatcherServletPath() {
    return () -> webMvcProperties.getServlet().getPath();
  }
}
```
